binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go

```diff
@@ -157,36 +164,77 @@ func getConsistencyFunc(c *client.ReadonlyClient) func(from, to uint64) ([][]byt
 	var pb api.ProofBundle
+	var fwMeta api.FirmwareMetadata
 
 	// Get the consistency proof for the bundle
 	dc, err := dev.DeviceCheckpoint()
 	if err != nil {
-		return pb, fmt.Errorf("failed to fetch the device checkpoint: %w", err)
+		return pb, fwMeta, fmt.Errorf("failed to fetch the device checkpoint: %w", err)
 	}
 
 	cpFunc := getConsistencyFunc(c)
 	fwHash := sha512.Sum512(up.FirmwareImage)
-	pb, err = verify.BundleForUpdate(up.ProofBundle, fwHash[:], dc, cpFunc)
+	pb, fwMeta, err = verify.BundleForUpdate(up.ProofBundle, fwHash[:], dc, cpFunc)
 	if err != nil {
-		return pb, fmt.Errorf("failed to verify proof bundle: %w", err)
+		return pb, fwMeta, fmt.Errorf("failed to verify proof bundle: %w", err)
 	}
-	return pb, nil
+	return pb, fwMeta, nil
 }
 
-// verifyWitness checks that an update package is consistent with witness
-func verifyWitness(c *client.ReadonlyClient, pb api.ProofBundle, wc client.WitnessClient) error {
+func verifyWitness(c *client.ReadonlyClient, pb api.ProofBundle, witnessURL string) error {
+	wURL, err := url.Parse(witnessURL)
+	if err != nil {
+		return fmt.Errorf("witness_url is invalid: %w", err)
+	}
+	wc := client.WitnessClient{URL: wURL}
+
 	wcp, err := wc.GetWitnessCheckpoint()
 	if err != nil {
 		return fmt.Errorf("failed to fetch the witness checkpoint: %w", err)
 	}
+	if err := verifyRemote(c, pb, *wcp); err != nil {
+		return fmt.Errorf("failed to verify update with witness: %w", err)
+	}
+	return nil
+}
+
+func verifyMap(c *client.ReadonlyClient, pb api.ProofBundle, fwMeta api.FirmwareMetadata, mapURL string) error {
+	mc := client.MapClient{}
+	mcp, err := mc.MapCheckpoint()
+	if err != nil {
+		return fmt.Errorf("failed to get map root: %w", err)
+	}
+	// TODO(mhutchinson): check consistency with the largest checkpoint found thus far
+	// in order to detect a class of fork; it could be that the checkpoint in the update
+	// is consistent with the map and the witness, but the map and the witness aren't
+	// consistent with each other.
+	if err := verifyRemote(c, pb, mcp.LogCheckpoint); err != nil {
+		return fmt.Errorf("failed to verify update with map checkpoint: %w", err)
+	}
+
+	// TODO(mhutchinson): Check the inclusion proof and use the values returned by the map.
+	afw, _, err := mc.Aggregation(mcp, pb.InclusionProof.LeafIndex)
+	if err != nil {
+		return fmt.Errorf("failed to get map value for %q: %w", pb.InclusionProof.LeafIndex, err)
+	}
+	if !reflect.DeepEqual(fwMeta, *afw.Firmware) {
+		return fmt.Errorf("got aggregated response for %q, but expected %q", afw.Firmware, fwMeta)
+	}
+	if !afw.Good {
+		return errors.New("firmware is marked as bad")
+	}
+	return nil
+}
 
-	if wcp.TreeSize == 0 {
-		return fmt.Errorf("No witness checkpoint to verify")
+// verifyRemote checks that an update package is consistent with remotely fetched Checkpoint.
+func verifyRemote(c *client.ReadonlyClient, pb api.ProofBundle, rcp api.LogCheckpoint) error {
```

Comment:
is this really `verifyConsistency` (same as the `verify` package)(?)

Reply:
Inlined this method out of existence.